cli/compose/template/template.go

```diff
@@ -47,49 +59,22 @@ func Substitute(template string, mapping Mapping) (string, error) {
 			substitution = groups["braced"]
 		}
 
-		switch {
-
-		case substitution == "":
+		if substitution == "" {
 			err = &InvalidTemplateError{Template: template}
 			return ""
+		}
 
-		// Soft default (fall back if unset or empty)
-		case strings.Contains(substitution, ":-"):
-			name, defaultValue := partition(substitution, ":-")
-			value, ok := mapping(name)
-			if !ok || value == "" {
-				return defaultValue
-			}
-			return value
-
-		// Hard default (fall back if-and-only-if empty)
-		case strings.Contains(substitution, "-"):
-			name, defaultValue := partition(substitution, "-")
-			value, ok := mapping(name)
-			if !ok {
-				return defaultValue
-			}
-			return value
-
-		case strings.Contains(substitution, ":?"):
-			name, errorMessage := partition(substitution, ":?")
-			value, ok := mapping(name)
-			if !ok || value == "" {
-				err = &InvalidTemplateError{
-					Template: fmt.Sprintf("required variable %s is missing a value: %s", name, errorMessage),
-				}
+		for _, f := range subsFuncs {
+			var (
+				value   string
+				applied bool
+			)
+			value, applied, err = f(substitution, mapping)
+			if err != nil {
```

Comment:
Is the error itself discarded?

Reply:
@thaJeztah it's not if it's the last one to be applied (i.e. if one function doesn't apply, or apply with error, it doesn't necessarly mean others wont).